### PR TITLE
[ML] Upgrade to Visual Studio 2019 for Windows compilation

### LINKED ISSUE #1015: [ML] Upgrade compilers so that we can use C++17 in the 8.x timeframe
Currently in our codebase we are able to use C++11 in the 6.x branches and C++14 in the 7.x branches.

It would be good if we can use C++17 in the 8.x branches (once they are eventually created).  Enabling this will require compiler upgrades on our release build environments for all platforms:

- [x] macOS - requires Xcode 10 or above - #867
- [x] Linux - requires gcc 7 (partial) or 9 (complete) - #1170
- [ ] Windows - requires Visual Studio 2019 - #1352

When the compilers are upgraded we will also take the opportunity to upgrade libxml2 from version 2.9.4 to version 2.9.7 with minimal extra effort.

### PR DESCRIPTION
**Pull request by droberts195**
We will build version 8.x of the ML C++ using Visual Studio 2019
rather than Visual Studio 2017.  This will eventually facilitate
use of C++17 (after separate PRs to upgrade the Mac and Linux
compilers have also been merged).

Since this change involves rebuilding all dependencies, the
libxml2 dependency is also upgraded from version 2.9.4 to
version 2.9.7.

Closes #1015

### CODE REVIEW

**Comment by droberts195** [comment]
This is nowhere near complete, but I wanted to create a branch to iterate on and thought I'd raise the PR too for increased visibility.

**Comment by droberts195** [comment]
The CI of this PR will always fail, as the Windows Jenkins worker it uses doesn't have VS2019 installed.  However, I tried a build on the internal London office Jenkins using a manually set up Windows worker and, running against the 12th commit of this PR, it succeeded!  For those who have London office VPN access see http://192.168.62.148:4343/job/MyBranchBuild-VS2019/3/

Remaining tasks before merging this PR are:

1. Merge master into the PR branch after #843 and #844 have been merged to master
2. Get Visual Studio 2019 added to the ML CI and release manager Windows VMs
3. Add a changelog entry
4. Get the PR reviewed

**Comment by edsavage** [comment]
> For those who have London office VPN access see http://192.168.62.148:4343/job/MyBranchBuild-VS2019/3/

Screenshot 2019-11-22 at 14 48 24

confirmed ;-)

**Comment by droberts195** [comment]
retest

**Comment by droberts195** [comment]
> 2. Get Visual Studio 2019 added to the ML CI and release manager Windows VMs

The first half of this is done, i.e. the CI VMs, hence we now have a green build. 🎉 

However, the second half, the release manager VM, is not done.  Hence this PR must not yet be merged as it would break the release builds. 😞 

It's ready for review though.

File: 3rd_party/3rd_party.sh [hunk 1]
```diff
@@ -88,13 +88,13 @@ case `uname` in
...
+        # Read VCBASE from environment if defined, otherwise default to VS Professional 2019
+        DEFAULTVCBASE=`cd /$LOCAL_DRIVE && cygpath -m -s "Program Files (x86)/Microsoft Visual Studio/2019/Professional"`
         VCBASE=${VCBASE:-$DEFAULTVCBASE}
         VCVER=`ls -1 /$LOCAL_DRIVE/$VCBASE/VC/Redist/MSVC | tail -1`
-        STL_LOCATION=/$LOCAL_DRIVE/$VCBASE/VC/Redist/MSVC/$VCVER/x64/Microsoft.VC141.CRT
-        STL_PREFIX=
-        STL_EXTENSION=140.dll
+        STL_LOCATION=/$LOCAL_DRIVE/$VCBASE/VC/Redist/MSVC/$VCVER/x64/Microsoft.VC142.CRT
+        STL_PATTERN=140
+        STL_EXTENSION=.dll
```
Comment (tveasey) [inline 3rd_party/3rd_party.sh]: 👍

**Comment by droberts195** [comment]
> * If we're upgrading libxml don't we also need to linux.md and Dockerfile.

When gcc 9.3 is released I'll do a separate PR that upgrades the Linux build environment to support C++17.  I'll upgrade libxml2 in this, so we're using the same version across all platforms in each branch.

**Comment by droberts195** [comment]
> * `CDetachedProcessSpawner_Windows.cc` references Visual Studio 2017. I wonder whether this comment still applies.

I adjusted this comment because on reading https://github.com/rprichard/win32-console-docs I could not see why the behaviour described as affecting Visual Studio 2017 wouldn't be the same on all versions of Windows (at least from Vista onwards, and we never supported anything earlier than that, even in 2011).  The more interesting thing is how it ever worked with `DETACHED_PROCESS` under Visual Studio 2013.  Possibly it didn't in general, but we didn't notice due to the bug in Windows 7 described in https://github.com/rprichard/win32-console-docs#windows-7-inheritability-win7inh.  Windows 7 is the same underlying Windows version as Windows Server 2008r2 (Windows NT 6.1), and we probably tested on Windows Server 2008r2 at the time we first switched to Visual Studio 2013.

**Pull request merged**